pkg/apis/seedmanagement/v1alpha1/defaults.go

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// SetDefaults_ManagedSeed sets default values for ManagedSeed objects.
+func SetDefaults_ManagedSeed(obj *ManagedSeed) {
+	switch {
+	case obj.Spec.SeedTemplate != nil:
+		setDefaultsSeedSpec(&obj.Spec.SeedTemplate.Spec, obj.Name, obj.Namespace, true)
+	case obj.Spec.Gardenlet != nil:
+		setDefaultsGardenlet(obj.Spec.Gardenlet, obj.Name, obj.Namespace)
+	}
+}
+
+// SetDefaults_GardenletDeployment sets default values for GardenletDeployment objects.
+func SetDefaults_GardenletDeployment(obj *GardenletDeployment) {
+	// Set default replica count
+	if obj.ReplicaCount == nil {
+		obj.ReplicaCount = pointer.Int32Ptr(1)
+	}
+
+	// Set default revision history limit
+	if obj.RevisionHistoryLimit == nil {
+		obj.RevisionHistoryLimit = pointer.Int32Ptr(1)
+	}
+
+	// Set default image
+	if obj.Image == nil {
+		obj.Image = &Image{}
+	}
+
+	// Set default VPA
+	if obj.VPA == nil {
+		obj.VPA = pointer.BoolPtr(true)
+	}
+}
+
+// SetDefaults_Image sets default values for Image objects.
+func SetDefaults_Image(obj *Image) {
+	// Set default pull policy
+	if obj.PullPolicy == nil {
+		var pullPolicy corev1.PullPolicy
+		if obj.Tag != nil && *obj.Tag == "latest" {
+			pullPolicy = corev1.PullAlways
+		} else {
+			pullPolicy = corev1.PullIfNotPresent
+		}
+		obj.PullPolicy = &pullPolicy
+	}
+}
+
+func setDefaultsGardenlet(obj *Gardenlet, name, namespace string) {
+	// Set deployment defaults
+	if obj.Deployment == nil {
+		obj.Deployment = &GardenletDeployment{}
+	}
+
+	// Decode / initialize gardenlet config
+	var gardenletConfig *configv1alpha1.GardenletConfiguration
+	var err error
+	if obj.Config != nil {
+		// Decode gardenlet config to an external version
+		// Without defaults, since we don't want to set gardenlet config defaults in the resource at this point
+		gardenletConfig, err = helper.DecodeGardenletConfigExternal(obj.Config, false)
+		if err != nil {
+			return
+		}
+	} else {
+		gardenletConfig = &configv1alpha1.GardenletConfiguration{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: configv1alpha1.SchemeGroupVersion.String(),
+				Kind:       "GardenletConfiguration",
+			},
+		}
+	}
+
+	// Set gardenlet config defaults
+	setDefaultsGardenletConfiguration(gardenletConfig, name, namespace)
+
+	// Set gardenlet config back to obj.Config
+	gardenletConfigJSON, _ := json.Marshal(gardenletConfig)
+	obj.Config = &runtime.RawExtension{Raw: gardenletConfigJSON}
```

Comment:
It shouldn't be necessary to convert to JSON by hand. the `RawExtension` type takes care of correct marshalling, see https://github.com/kubernetes/apimachinery/blob/0e9f9cff521d6ebb5aaee87ae496b501bc27e974/pkg/runtime/extension.go#L37
```suggestion
	obj.Config = &runtime.RawExtension{Object: gardenletConfig}
```

Reply:
OK